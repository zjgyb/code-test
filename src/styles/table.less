/*
 * @Author: gyb
 * @Date: 2021-10-17
 * @Description: 表格样式
 */

.sf-table {
  width: 100%;
  text-align: left;
  border-collapse: separate;
  border-spacing: 0;
  table-layout: fixed;

  &-thead &-cell {
    padding: 16px;
    background-color: #fafafa;
  }

  &-sorter {
    display: inline-flex;
    flex-direction: column;
    height: 20px;
    color: #bfbfbf;

    &-down {
      transform: translateY(-10px);
    }

    &--active {
      color: var(--active-color);
    }
  }

  &-placeholder {
    text-align: center;
  }

  &-cell {
    padding: 0 16px;

    &-hbox {
      display: flex;
      justify-content: space-between;
      align-items: center;
    }

    &-title {
      margin-right: 8px;
      flex: 1;
      min-width: 0;
    }

    &-selection {
      display: flex;
      justify-content: center;
    }

    &--sorable {
      cursor: pointer;
    }

    &--normal {
      height: 40px;
    }

    &--sm {
      height: 32px;
    }

    &--lg {
      height: 48px;
    }
  }

  &-pagination {
    display: flex;
    justify-content: end;
    margin-top: 10px;
  }
}
