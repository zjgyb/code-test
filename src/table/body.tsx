/*
 * @Author: gyb
 * @Date: 2021-10-18
 * @Description: 表格主体相关界面
 */

import type { TableBodyOptions } from './types';

import { SfCheckbox } from '../form/checkbox';

export function renderBody(options: TableBodyOptions) {
	const { context, props, tableData, onChangeCheckbox, tdClasses, tdStyles } = options;	

	return (
		tableData.value.map(record => {
			return <tr>
				{ props.selectable && 
					<td class={tdClasses.value} style={tdStyles.value}>
						<div class="sf-table-cell-selection">
							<SfCheckbox checked={record.selected} onChange={() => onChangeCheckbox(record)} />
						</div>
					</td> 
				}
				{
					props.columns.map(column => {

						// 判断是否存在slots
						const vnodes = context.slots?.body?.({column, record});					
						const hasVNode = vnodes?.some(vnode => vnode.text || vnode.tag);
						return <td class={tdClasses.value} style={tdStyles.value} colspan={1}>
							{ hasVNode ? vnodes : record[column.key] }
						</td>
					})
				}
			</tr>
		})
	)
}