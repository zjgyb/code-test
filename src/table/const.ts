/*
 * @Author: gyb
 * @Date: 2021-10-18
 * @Description: 表格组件默认配置项
 */

// 表格分页配置默认值
export const pagination = {
	pageSize: 10,
	current: 1,
	total: 0
}

// 排序相关
export enum Order {
	desc = 'desc',
	asc = 'asc'
}

// 排序相关，用于改变排序值
export const orderMap = new Map<Order, Order | ''>([
	[Order.asc, Order.desc],
	[Order.desc, '']
]);

// 表格高度相关
export enum TableSize {
	small = 'sm',
	normal = 'normal',
	large = 'lg'
}