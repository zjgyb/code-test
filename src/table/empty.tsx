/*
 * @Author: gyb
 * @Date: 2021-10-18
 * @Description: 空状态相关界面
 */

import type { SetupContext } from '@vue/composition-api';
import type { TablePublicProps } from './types';

export function renderEmpty(props: TablePublicProps, context: SetupContext) {
	return (
		<tr class="sf-table-placeholder">
			<td class="sf-table-cell" colspan={props.columns.length}>
				{context.slots?.empty?.() || props.emptyText}
			</td>
		</tr>
	)
}