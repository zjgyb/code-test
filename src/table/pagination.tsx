/*
 * @Author: gyb
 * @Date: 2021-10-18
 * @Description: 表格分页相关代码
 */

import type { TablePaginationOptions } from './types';

import { SfPagination } from '../pagination';

export function renderPagination(options: TablePaginationOptions) {
    const { pagination, props, onChangePager } = options;
    
    if (props.pagination === false) {
        return null;
    }
    
    return <SfPagination class="sf-table-pagination" current={pagination.current} total={pagination.total} onChange={onChangePager} />
}