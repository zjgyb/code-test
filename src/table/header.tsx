/*
 * @Author: gyb
 * @Date: 2021-10-18
 * @Description: 表格表头
 */

import type { TableHeaderOptions } from './types';

import { SfCheckbox } from '../form/checkbox';
import { Order } from './const';

export function renderHeader(options: TableHeaderOptions) {
	const {
		sortField,
		context,
		props,
		checked,
		isHalfChecked,
		onChangeSelections,
		onChangeSort } = options;

	if (!props.hasHeader) {
		return null;
	}

	return (
		<thead class="sf-table-thead">
			<tr>
				{props.selectable &&
					<th class="sf-table-cell" colspan="1" style={{ width: '40px' }}>
						<div class="sf-table-cell-selection">
							<SfCheckbox checked={checked.value} onChange={onChangeSelections} isHalf={isHalfChecked.value} />
						</div>
					</th>
				}
				{
					props.columns.map(column => {
						const isSameKey = sortField.key === column.key;
						const isAscActive = isSameKey && sortField.order === Order.asc;
						const isDescActive = isSameKey && sortField.order === Order.desc;
						const vnodes = context.slots?.header?.({ column });
						const hasVNode = vnodes?.some(vnode => vnode.text || vnode.tag);

						return (
							<th class={{ 'sf-table-cell': true, 'sf-table-cell--sorable': column.sortable }} colspan="1" onClick={() => onChangeSort(column)}>
								<div class="sf-table-cell-hbox">
									{hasVNode ? vnodes : <span class="sf-table-cell-title ellipsis" title={column.title}>{column.title}</span>}
									{column.sortable &&
										<span class="sf-table-sorter">
											<i class={`fas fa-sort-up sf-table-sorter-up ${isAscActive ? 'sf-table-sorter--active' : ''}`}></i>
											<i class={`fas fa-sort-down sf-table-sorter-down ${isDescActive ? 'sf-table-sorter--active' : ''}`}></i>
										</span>
									}
								</div>
							</th>
						)
					})
				}
			</tr>
		</thead>
	)
}