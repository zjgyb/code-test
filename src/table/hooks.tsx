/*
 * @Author: gyb
 * @Date: 2021-10-30
 * @Description: 表格hooks
 */

import type { Pagination, SortKeys, TableColumn, TableRecord } from './types';
import { reactive, ref, Ref, computed, watch, getCurrentInstance, ComputedRef, watchEffect } from '@vue/composition-api';
import { cloneDeep, merge, sortBy, isNumber, isString } from 'lodash-es';
import { Order, pagination, orderMap, TableSize } from './const';

// 分页相关
export function useSort() {
	const sortField = reactive<Partial<SortKeys>>({ key: '', order: '' });

	/**
	 * 排序函数，一般情况下先顺序再倒序
	 * @param column 表格列参数
	 */
	const onChangeSort = (column: TableColumn) => {
		sortField.key = column.key;
		sortField.order = sortField.order ? orderMap.get(sortField.order) : Order.asc;
	}

	return {
		sortField,
		onChangeSort
	}
}

// 监听传入pagination变化
export function usePager(options: Ref<Pagination | false>, total: ComputedRef<number>) {
	const paginationOptions = reactive(merge({}, pagination, { total: total.value }));

	/**
	* 当表格页码改变时改变值
	* @param value 页码
	*/
	const onChangePager = (value: number) => {
		paginationOptions.current = value;
	}

	watchEffect(() => {
		paginationOptions.total = (options.value ? options.value.total : null) ?? total.value;
	});

	// 改变分页配置
	watch(options, (newVal) => {     
		merge(paginationOptions, newVal || {});
	}, { deep: true });

	return {
		paginationOptions,
		onChangePager
	}
}

// 与数据管理
export function useData(data: Ref<AnyObject[]>, sortField: Partial<SortKeys>, pagination: Pagination, isShowPager: ComputedRef<boolean>) {
	const sourceData = computed<TableRecord[]>(() => {
		return cloneDeep(data.value).map(item => {
			return { ...item, selected: item.selected ?? false };
		});
	});
	const vm = getCurrentInstance();

	// 排序后数据
	const sortData = computed(() => {
		const sortByData = sortBy(sourceData.value, [sortField.key]);
		sortField.order === Order.desc && sortByData.reverse();
		return sortField.order ? sortByData : sourceData.value;
	});

	// 判读是否需要截取数据，如果截取数据不为空，则截取，否则直接用源数据
	const sliceData = computed(() => {
		if (!isShowPager.value) {
			return sourceData.value;
		}

		const startPage = (pagination.current - 1) * pagination.pageSize;
		let sliceData = sortData.value?.slice(startPage, startPage + pagination.pageSize);
		return sliceData.length ? sliceData : sortData.value;
	});

	const tableData = ref(sliceData.value);

	// 页码或者排序变化时向外抛出事件
	watch(isShowPager.value ? [sortField, pagination] : [sortField], () => {
		vm?.emit('change', pagination, sortField);
	}, { deep: true });

	watch(sliceData, () => {
		tableData.value = sliceData.value;
	});

	return {
		tableData,
		sortData
	}
}

// 勾选相关
export function useChecked(tableData: Ref<TableRecord[]>, allData: Ref<TableRecord[]>) {
	const checked = ref(false);
	const vm = getCurrentInstance();

	const selectedData = computed(() => {
		return allData.value.filter(item => item.selected);
	});

	// 是否处于部分勾选状态
	const isHalfChecked = computed(() => {
		const selectedLen = selectedData.value.length;
		return selectedLen > 0 && tableData.value.length > selectedLen;
	});

	/**
	 * 改变勾选状态
	 * @param isChecked 勾选状态
	 */
	const onChangeSelections = (isChecked: boolean) => {
		checked.value = isChecked;
		const selectedItems: TableRecord[] = [];
		tableData.value.forEach(item => {
			if (!item.selected) {
				selectedItems.push(item);
			}
			item.selected = isChecked;
		});
		vm?.emit('select', selectedItems, selectedData);
	}

	// 表格勾选状态改变
	const onChangeCheckbox = (record: TableRecord) => {
		record.selected = !record.selected;
		vm?.emit('select', [record], selectedData.value);
	}

	watch(tableData, () => {
		checked.value = tableData.value.every(item => item.selected);
	}, { immediate: true });

	return {
		checked,
		onChangeCheckbox,
		onChangeSelections,
		isHalfChecked
	}
}

// 样式相关
export function useClasses(rowHeight: Ref<TableSize | number>) {

	// 如果传入的字符串，则根据传入的值渲染类名
	const tdClasses = computed(() => {
		const isTableSize = isString(rowHeight.value) && Object.values(TableSize).includes(rowHeight.value);
		const size = isTableSize ? rowHeight.value : TableSize.normal;
		return {
			'sf-table-cell': true,
			[`sf-table-cell--${size}`]: true
		}
	});

	// 如果传入的是number，则使用style进行渲染
	const tdStyles = computed(() => {
		return isNumber(rowHeight.value) ? {
			height: rowHeight.value + 'px'
		} : {}
	});

	return {
		tdClasses,
		tdStyles
	};
}
