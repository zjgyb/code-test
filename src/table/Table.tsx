/*
 * @Author: gyb
 * @Date: 2021-10-17
 * @Description: 表格组件
 */
import type { TablePublicProps } from './types';

import { computed, defineComponent, toRefs } from '@vue/composition-api';
import { tableProps } from './types';
import { renderHeader } from './header';
import { renderPagination } from './pagination';
import { renderEmpty } from './empty';
import { renderBody } from './body';
import { useClasses } from './composables/use-classes';
import { useSort } from './composables/use-sort';
import { usePager } from './composables/use-pager';
import { useChecked } from './composables/use-checked';
import { useData } from './composables/use-data';

export default defineComponent({
  name: 'Table',
  props: tableProps,
  setup(props: TablePublicProps, context) {
    const { rowHeight, data, pagination } = toRefs(props);
    const isShowPager = computed(() => !!pagination.value);
    const total = computed(() => data.value.length);
    const { sortField, onChangeSort } = useSort();
    const { paginationOptions, onChangePager } = usePager(pagination, total);
    const { tableData, sortData } = useData(data, sortField, paginationOptions, isShowPager);
    const { checked, onChangeSelections, onChangeCheckbox, isHalfChecked } = useChecked(tableData, sortData);
    const { tdClasses, tdStyles } = useClasses(rowHeight);

    return () => {
      const headerNode = renderHeader({ props, context, sortField, checked, onChangeSelections, isHalfChecked, onChangeSort });
      const emptyNode = renderEmpty(props, context);
      const bodyNode = renderBody({ props, context, tableData, tdClasses, tdStyles, onChangeCheckbox });
      const paginationNode = renderPagination({ props, context, pagination: paginationOptions, onChangePager });

      return (
        <div>
          <table class="sf-table">
            {/* <colgroup>
              {props.columns.map(() => <col />)}
            </colgroup> */}
            {headerNode}
            <tbody class="sf-table-tbody">
              {!props.data.length ? emptyNode : bodyNode}
            </tbody>
          </table>
          { props.pagination !== false && paginationNode}
        </div>
      )
    }
  },
});
