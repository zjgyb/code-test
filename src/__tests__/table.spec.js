import { mount } from '@vue/test-utils'
import { TestTable } from '../table'

export const columns = [{
        title: 'name1111111111111111111111111111111111111111111111111',
        dataIndex: 'name',
        key: 'name',
        sortable: true
    },
    {
        title: 'Age',
        dataIndex: 'age',
        key: 'age',
    },
    {
        title: 'Address',
        dataIndex: 'address',
        key: 'address',
    }
];

let testData = [];
for (let i = 0; i < 100; i++) {
    testData.push({
        name: 'John Brown' + i,
        age: i,
        address: `New York No. ${i} Lake Park`
    })
}
export const data = testData;

describe('Table', () => {
    const TableMount = options => mount(TestTable, options)

    test('render', () => {
        const wrapper = TableMount()
        expect(wrapper.html()).toMatchSnapshot()
        expect(() => {
            wrapper.vm.$forceUpdate()
            wrapper.vm.$destroy()
        }).not.toThrow()
    });

    test('props empty data', () => {
        const wrapper = TableMount({
            propsData: {
                data: [],
                columns
            },
        })

        expect(wrapper.find('.sf-table-placeholder').exists()).toBeTruthy()
    })

    test('props emptyText', () => {
        const wrapper = TableMount({
            propsData: {
                data: [],
                columns,
                emptyText: 'No'
            },
        })
        expect(wrapper.find('.sf-table-placeholder .sf-table-cell').text()).toEqual('No')
    })

    test('sortable', async() => {
        const wrapper = TableMount({
            propsData: {
                data: [],
                columns,
            }
        });
        expect(wrapper.find('.sf-table-sorter').exists()).toBeTruthy();
        const sortHeader = wrapper.find('.sf-table-thead .sf-table-cell');
        sortHeader.trigger('click');
        expect(wrapper.find('.fa-sort-up').exists()).toBeTruthy();
        await wrapper.vm.$nextTick();
        expect(wrapper.find('.fa-sort-up.sf-table-sorter--active').exists()).toBeTruthy();
        sortHeader.trigger('click');
        await wrapper.vm.$nextTick();
        expect(wrapper.find('.sf-table-sorter .fa-sort-down.sf-table-sorter--active').exists()).toBeTruthy();
        await wrapper.vm.$nextTick();
        sortHeader.trigger('click');
        await wrapper.vm.$nextTick();
        expect(wrapper.find('.sf-table-sorter .sf-table-sorter--active').exists()).toBeFalsy();
    });

    test('pagination', async() => {
        let testData = [];
        for (let i = 0; i < 100; i++) {
            testData.push({
                name: 'John Brown' + i,
                age: i,
                address: `New York No. ${i} Lake Park`
            })
        }
        const wrapper = TableMount({
            propsData: {
                data: testData,
                columns,
                pagination: {
                    current: 1
                }
            }
        });
        expect(wrapper.find('.sf-table-pagination').exists()).toBeTruthy()
        const nextBtns = wrapper.findAll('.sf-pagination-item')
        nextBtns.at(nextBtns.length - 1).trigger('click');
        await wrapper.vm.$nextTick();
        expect(wrapper.find('.sf-pagination-item--active').text()).toEqual('2');
        wrapper.setProps({ pagination: { current: 3 } });
        await wrapper.vm.$nextTick();
        expect(wrapper.find('.sf-pagination-item--active').text()).toEqual('3');

        wrapper.setProps({ pagination: false });
        await wrapper.vm.$nextTick();
        expect(wrapper.find('.sf-table-pagination').exists()).toBeFalsy();
    });

    test('pagination empty', async() => {
        let testData = [];
        for (let i = 0; i < 100; i++) {
            testData.push({
                name: 'John Brown' + i,
                age: i,
                address: `New York No. ${i} Lake Park`
            })
        }
        const wrapper = TableMount({
            propsData: {
                data: testData,
                columns,
                pagination: false
            }
        });
        expect(wrapper.find('.sf-table-pagination').exists()).toBeFalsy()
        expect(wrapper.findAll('tbody tr').length).toBe(testData.length);
    });

    test('slots empty', () => {
        const emptySlots = {
            template: `<span class="test-empty">HaHaHa</span>`
        }
        const wrapper = TableMount({
            propsData: {
                data: [],
                columns,
            },
            slots: {
                empty: emptySlots
            }
        });
        expect(wrapper.html()).toContain(emptySlots.template)
        expect(wrapper.find('.test-empty').exists()).toBeTruthy()
        expect(wrapper.find('.test-empty').text()).toEqual('HaHaHa')
    })

    test('slots header', () => {
        const headerSlots = `
              <template v-if="column.key === 'name'">
                <span class="test-header">
                  test
                </span>
              </template>
          `;

        const wrapper = TableMount({
            propsData: {
                data: [],
                columns,
            },
            scopedSlots: {
                header: `<template slot-scope="{column}">${headerSlots}</template>`
            }
        });

        expect(wrapper.find('.test-header').exists()).toBeTruthy();
        expect(wrapper.find('.test-header').text()).toEqual('test');
    });

    test('table header hide', () => {
        const wrapper = TableMount({
            propsData: {
                data,
                columns,
                hasHeader: false
            },
        });
        expect(wrapper.find('.sf-table-thead').exists()).toBeFalsy();
    });

    test('slots body', async() => {
        const bodySlots = `
              <template v-if="column.key === 'name'">
                <span class="test-body">
                  test
                </span>
              </template>
          `;

        const wrapper = TableMount({
            propsData: {
                data,
                columns,
            },
            scopedSlots: {
                body: `<template slot-scope="{column}">${bodySlots}</template>`
            }
        });

        expect(wrapper.find('.test-body').exists()).toBeTruthy();
        expect(wrapper.find('.test-body').text()).toEqual('test');
    });

    test('table selectable', async() => {
        const wrapper = TableMount({
            propsData: {
                data,
                columns,
                selectable: true
            },
        });

        const bodyCheckbox = wrapper.find('tbody .sf-checkbox');
        const headerCheckbox = wrapper.find('thead .sf-checkbox');
        expect(bodyCheckbox.exists()).toBeTruthy();
        const bodyCheckboxInput = bodyCheckbox.find('input');
        expect(bodyCheckboxInput.exists()).toBeTruthy();
        await bodyCheckboxInput.setChecked();
        expect(bodyCheckboxInput.element.checked).toBeTruthy();
        expect(wrapper.emitted().select).toBeTruthy();
        expect(wrapper.emitted().select[0].length).toBe(2);
        expect(wrapper.emitted().select[0][0]).toEqual([{...data[0], selected: true }]);
        const headerCheckboxInput = headerCheckbox.find('input');
        expect(headerCheckboxInput.exists()).toBeTruthy();
        await headerCheckboxInput.setChecked();
        expect(headerCheckboxInput.element.checked).toBeTruthy();
    });

    test('table rowHeight', () => {
        const wrapper = TableMount({
            propsData: {
                data,
                columns,
                rowHeight: 64
            },
        });
        const td = wrapper.find('tbody td');
        expect(td.attributes().style).toContain('64px');
    });
})