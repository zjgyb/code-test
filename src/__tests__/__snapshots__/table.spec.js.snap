// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render 1`] = `
"<div>
  <table class=\\"sf-table\\">
    <colgroup></colgroup>
    <thead class=\\"sf-table-thead\\">
      <tr></tr>
    </thead>
    <tbody class=\\"sf-table-tbody\\">
      <tr class=\\"sf-table-placeholder\\">
        <td colspan=\\"0\\" class=\\"sf-table-cell\\">No data</td>
      </tr>
    </tbody>
  </table>
  <ul class=\\"sf-pagination sf-table-pagination\\">
    <li title=\\"上一页\\" class=\\"sf-pagination-item sf-pagination-item--disabled\\">&lt;</li>
    <li title=\\"1\\" class=\\"sf-pagination-item sf-pagination-item--active\\">1</li>
    <li title=\\"下一页\\" class=\\"sf-pagination-item\\">&gt;</li>
  </ul>
</div>"
`;
